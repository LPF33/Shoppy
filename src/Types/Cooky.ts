// Types for Cooky View & Components

export interface IMeal {
  key?: string;
  meal?: string;
  startDate: string;
  endDate: string;
}

export interface IComputedMealList {
  key: string;
  meal: string;
  startDate: string;
  endDate: string;
  color?: string;
  startNum?: number;
  endNum?: number;
}

export interface INewMeal {
  startDate: Date;
  startNum: number;
}

export interface ICookyAddMealData {
  today: Date;
  startDate: null | string;
  endDate: null | string;
  meal: string;
  existingKey: false | string;
}

export interface ICookyCalendarData {
  weekday: ["Mo", "Di", "Mi", "Do", "Fr", "Sa", "So"];
  today: Date;
}
