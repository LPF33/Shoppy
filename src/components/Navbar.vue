<template>
  <nav v-if="isLoggedIn">
    <router-link :to="{ name: 'Main' }">Main</router-link>
    <router-link :to="{ name: 'Shoppy' }">Shoppy</router-link>
    <router-link :to="{ name: 'Cook' }">Cooky</router-link>
  </nav>
  <nav v-else>
    <router-link :to="{ name: 'Login' }">Login</router-link>
  </nav>
</template>

<script>
export default {
  name: "Navbar",
  computed: {
    isLoggedIn: function () {
      return this.$store.state.isAuthenticated;
    },
  },
};
</script>

<style>
nav {
  width: 100%;
  padding: 10px;
  display: flex;
  justify-content: center;
  background: #232323;
}

nav a {
  font-weight: bold;
  color: #2c3e50;
  border: 2px solid rgb(184, 184, 184);
  padding: 10px;
  width: 150px;
  text-align: center;
  margin: 0 10px;
  color: #ffeda3;
  text-decoration: none;
}

nav a.router-link-exact-active {
  color: #39a9cb;
}
</style>
