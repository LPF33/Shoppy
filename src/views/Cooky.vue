<template>
  <main>
    <h1>👩‍🍳 👨‍🍳</h1>
    <h1>Wann kochen wir was?</h1>
    <CookyCalendar
      v-show="!modalVisible"
      :meals="mealList"
      @open-modal="openCookyAddMeal"
    />
    <CookyAddMeal
      v-show="modalVisible"
      @click.self="modalVisible = false"
      @close="modalVisible = false"
      :data-object="mealData"
    />
  </main>
</template>

<script>
import CookyCalendar from "@/components/CookyCalendar.vue";
import CookyAddMeal from "@/components/CookyAddMeal.vue";

export default {
  name: "Cooky",
  components: {
    CookyCalendar,
    CookyAddMeal,
  },
  data() {
    return {
      modalVisible: false,
      mealData: null,
    };
  },
  methods: {
    getMealDay(date) {
      date = new Date(date);
      return date.getDate();
    },
    openCookyAddMeal(meal) {
      this.mealData = meal;
      this.modalVisible = true;
    },
  },
  computed: {
    mealList() {
      return this.$store.state.mealList.map((item) => {
        const randomNum = Math.floor(Math.random() * (361 - 200) + 200);
        item.color = `hsl(${randomNum}, 80%, 70%)`;
        item.startNum = this.getMealDay(item.startDate);
        item.endNum = this.getMealDay(item.endDate);
        return item;
      });
    },
  },
};
</script>

<style scoped>
main {
  position: relative;
  background-color: #39a9cb;
  width: 100%;
  height: 100%;
  color: #232323;
  text-align: center;
}

h1 {
  margin-bottom: 10px;
}
</style>
