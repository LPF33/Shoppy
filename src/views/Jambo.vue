<template>
  <section>
    <h1>Jambo</h1>
    <h2 v-if="!counter">Start</h2>
    <h2 v-else>Spielzug: {{ counter }}</h2>
    <div v-if="visual">
      <div v-for="(item, index) in visual" :key="index" class="runde">
        <font-awesome-icon
          icon="fa-duotone fa-card-heart"
          style="
            --fa-secondary-opacity: 1;
            --fa-primary-color: rgb(255, 92, 92);
            --fa-secondary-color: white;
          "
        />
      </div>
    </div>
    <button @click="increment">
      <font-awesome-icon
        icon="fa-duotone fa-cards-blank"
        style="
          --fa-secondary-opacity: 1;
          --fa-primary-color: rgb(246, 90, 131);
          --fa-secondary-color: rgb(127, 183, 126);
        "
      />
    </button>
    <button @click="reset">
      <font-awesome-icon
        icon="fa-duotone fa-eraser"
        style="
          --fa-secondary-opacity: 1;
          --fa-primary-color: red;
          --fa-secondary-color: black;
        "
      />
    </button>
  </section>
</template>

<script lang="ts">
import { defineComponent } from "vue";

export default defineComponent({
  name: "Jambo",
  data() {
    return {
      counter: 0,
    };
  },
  methods: {
    increment() {
      if (this.counter >= 5) {
        this.reset();
        return;
      }
      this.counter++;
    },
    reset() {
      this.counter = 0;
    },
  },
  computed: {
    visual(): any[] | null {
      return this.counter > 0 ? new Array(this.counter) : null;
    },
  },
});
</script>

<style scoped>
section {
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  text-align: center;
  background-color: rgb(255, 232, 152);
  user-select: none;
}

h1,
h2 {
  margin: 10px 0;
}

button {
  display: block;
  font-size: 3rem;
  height: 6rem;
  width: 6rem;
  border: 1px solid grey;
  border-radius: 50%;
  background-color: transparent;
  line-height: 1.5;
  padding: 10px;
  cursor: pointer;
  margin-top: 2rem;
  user-select: none;
  box-shadow: 0 0 8px black;
}

.runde {
  font-size: 3rem;
}
</style>
