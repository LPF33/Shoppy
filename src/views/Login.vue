<template>
  <div v-if="error" class="error">{{ errorMessage }}</div>
  <form @submit.prevent="submit">
    <h1>Login</h1>
    <input
      type="email"
      required
      placeholder="email"
      v-model="email"
      ref="input"
    />
    <input type="password" required placeholder="password" v-model="password" />
    <button type="submit">Submit</button>
  </form>
</template>

<script>
export default {
  name: "Login",
  data() {
    return {
      email: "",
      password: "",
    };
  },
  methods: {
    submit() {
      this.$store.dispatch("signIn", {
        email: this.email,
        password: this.password,
      });
    },
  },
  computed: {
    error() {
      return this.$store.state.error;
    },
    errorMessage() {
      return this.$store.state.errorMessage;
    },
  },
  mounted() {
    this.$refs.input.focus();
  },
};
</script>

<style scoped>
h1 {
  padding-bottom: 10vh;
}

form {
  height: 100%;
  max-width: 500px;
  margin: 0 auto;
  display: flex;
  flex-direction: column;
  justify-content: center;
  text-align: center;
}

button[type="submit"] {
  background-color: #39a9cb;
  color: white;
  padding: 12px 20px;
  cursor: pointer;
  border-radius: 30px;
  margin: 20px 10%;
}
button[type="submit"]:hover {
  background-color: #2940d3;
}
input {
  margin: 5px;
  box-shadow: 0 0 15px 4px rgba(0, 0, 0, 0.06);
  padding: 10px;
  border-radius: 30px;
  margin: 20px 10%;
}
</style>
