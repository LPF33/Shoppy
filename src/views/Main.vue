<template>
  <section>
    <img :src="`${publicPath}img/wedding.JPG`" alt="wedding-photo" />
    <div v-if="schaetzchen" v-html="greeting" class="header"></div>
    <div v-else class="emoji">ðŸ¦‹</div>
    <aside>{{ wedding }} Tage bis zur Hochzeit</aside>
    <footer>
      <router-link :to="{ name: 'Budget' }" class="buttons">ðŸ’°</router-link>
      <router-link :to="{ name: 'Jambo' }" class="buttons">ðŸŽ²</router-link>
      <button @click="userLogout" class="buttons">ðŸ”’</button>
    </footer>
  </section>
</template>

<script>
export default {
  name: "Main",
  data() {
    return {
      publicPath: process.env.BASE_URL,
    };
  },
  methods: {
    userLogout() {
      this.$store.commit("unsubscribeFirebase");
      this.$store.commit("logout");
    },
  },
  computed: {
    wedding() {
      const weddingDate = new Date(2021, 8, 7);
      const now = new Date();
      return (
        Math.floor(
          (weddingDate.getTime() - now.getTime()) / (1000 * 3600 * 24)
        ) + 1
      );
    },
    schaetzchen() {
      return this.$store.state.user !== "lars@gmail.com";
    },
    greeting() {
      const now = new Date();
      const hours = now.getHours();
      if (5 <= hours && hours <= 11) {
        return "<h1>Guten Morgen ðŸŒž!</h1><h1>SchÃ¶nen Tag, meine Kleine â™¥</h1>";
      } else if (11 < hours && hours <= 14) {
        return "<h1>SchÃ¶nen Mittag ðŸŒ¼ðŸ¦‹</h1>";
      } else if (14 < hours && hours <= 17) {
        return "<h1>SchÃ¶nen Nachmittag â˜•</h1>";
      } else if (17 < hours && hours <= 22) {
        return "<h1>SchÃ¶nen Abend ðŸ¥°</h1><h1>mein kleines SchÃ¤tzchen ðŸ¥°</h1>";
      } else {
        return "<h1>Schlaf schÃ¶n ðŸ˜´</h1><h1> Gute Nacht, meine Kleine â™¥</h1>";
      }
    },
  },
};
</script>

<style scoped>
section {
  position: relative;
  background-color: #ffffff;
  height: 100%;
  color: #232323;
  font-weight: bold;
  display: flex;
  flex-direction: column;
  align-items: center;
  text-align: center;
  z-index: 1;
}

img {
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  max-width: 100vw;
  max-height: 100%;
  object-fit: contain;
  z-index: -1;
}

.header {
  padding: 20px;
}

aside {
  font-size: 1.5rem;
  padding: 20px;
  margin: auto 0;
  background-color: rgba(0, 0, 0, 0.507);
  border-radius: 10px;
  color: rgb(255, 255, 255);
}

footer {
  position: absolute;
  bottom: 10px;
  display: flex;
  justify-content: space-between;
  width: 100%;
  padding: 0 10px;
}

.buttons {
  background-color: transparent;
  font-size: 3rem;
  height: 6rem;
  width: 6rem;
  border: 1px solid grey;
  border-radius: 50%;
  line-height: 1.5;
  padding: 10px;
  cursor: pointer;
  user-select: none;
  box-shadow: 0 0 8px black;
  text-decoration: none;
}

.emoji {
  font-size: 3rem;
  padding: 10px;
  margin-top: 1rem;
  user-select: none;
}
</style>
