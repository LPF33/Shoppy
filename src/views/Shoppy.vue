<template>
  <section @click.self="setVisible">
    <h1>Shoppy</h1>
    <div>
      <div class="section-draggable" v-if="category === 'Supermarkt'">
        <VueDraggableNext
          v-model="supermarktItems"
          @change="onDrop"
          class="draggable-next"
        >
          <span v-for="it in supermarktItems" :key="it.key">üñê</span>
        </VueDraggableNext>
        <div>
          <ShoppingItem
            v-for="it in supermarktItems"
            :key="it.key"
            :item="it"
          />
        </div>
      </div>
      <div class="section-draggable" v-if="category === 'Drogerie'">
        <VueDraggableNext
          v-model="drogerieItems"
          @change="onDrop"
          class="draggable-next"
        >
          <span v-for="it in drogerieItems" :key="it.key">üñê</span>
        </VueDraggableNext>
        <div>
          <ShoppingItem v-for="it in drogerieItems" :key="it.key" :item="it" />
        </div>
      </div>
      <AddItem :index="index" :selected="category" />
    </div>
    <footer>
      <button
        @click="switchCategory('Supermarkt')"
        :class="{ active: category === 'Supermarkt' }"
      >
        üßÄ
      </button>
      <button>üé®</button>
      <button
        @click="switchCategory('Drogerie')"
        :class="{ active: category === 'Drogerie' }"
      >
        üß¥
      </button>
    </footer>
  </section>
</template>

<script>
import AddItem from "@/components/AddItem.vue";
import ShoppingItem from "@/components/ShoppingItem.vue";
import { shoppyFirestore } from "@/firebase/config";
import { VueDraggableNext } from "vue-draggable-next";

export default {
  name: "Shoppy",
  components: {
    AddItem,
    ShoppingItem,
    VueDraggableNext,
  },
  data() {
    return {
      supermarktItems: this.$store.state.supermarktItems,
      drogerieItems: this.$store.state.drogerieItems,
      category: "Supermarkt",
      visible: false,
      index: this.$store.state.shoppyIndex ?? 0,
      unwatch: null,
    };
  },
  methods: {
    setVisible() {
      this.visible = !this.visible;
    },
    switchCategory(category) {
      this.category = category;
    },
    onDrop(e) {
      const { oldIndex, newIndex, element: draggedItem } = e.moved;
      const itemsList =
        draggedItem.category === "Supermarkt"
          ? [...this.supermarktItems]
          : [...this.drogerieItems];
      let newDraggedIndex;
      if (newIndex === oldIndex) {
        return;
      }
      if (newIndex < oldIndex) {
        newDraggedIndex = itemsList[newIndex + 1].list_id;
        itemsList.forEach((item) => {
          if (
            item.list_id >= newDraggedIndex &&
            item.list_id !== draggedItem.list_id
          ) {
            item.list_id++;
            this.updateFirebase(item);
          }
        });
      } else {
        newDraggedIndex = itemsList[newIndex - 1].list_id;
        itemsList.forEach((item) => {
          if (
            item.list_id <= newDraggedIndex &&
            item.list_id !== draggedItem.list_id
          ) {
            item.list_id--;
            this.updateFirebase(item);
          }
        });
      }
      draggedItem.list_id = newDraggedIndex;
      this.updateFirebase(draggedItem);
    },
    async updateFirebase(item) {
      await shoppyFirestore
        .collection("shoppy")
        .doc(item.key)
        .update({ list_id: item.list_id });
    },
  },
  created() {
    this.unwatch = this.$store.watch(
      (state, getters) => getters.shoppyState,
      (newValue) => {
        this.supermarktItems = newValue.supermarktItems;
        this.drogerieItems = newValue.drogerieItems;
        this.index = newValue.shoppyIndex;
      }
    );
  },
  beforeUnmount() {
    if (typeof this.unwatch === "function") {
      this.unwatch();
    }
  },
};
</script>

<style scoped>
section {
  position: relative;
  background-color: #232323;
  width: 100%;
  height: 100%;
  color: #ffeda3;
  font-weight: bold;
  display: flex;
  flex-direction: column;
  align-items: center;
  text-align: center;
  user-select: none;
}

h1 {
  margin: 10px 0;
  height: 1.5em;
}

section > div {
  position: absolute;
  top: calc(2rem + 20px);
  bottom: 8rem;
  max-width: 500px;
  width: 90vw;
  overflow-y: auto;
}

.section-draggable {
  display: flex;
  overflow: hidden;
}

.section-draggable > div:nth-child(2) {
  flex-grow: 1;
  overflow-x: auto;
}

.draggable-next {
  display: flex;
  flex-direction: column;
  margin-right: 5px;
}

.draggable-next span {
  height: 40px;
  font-size: 2rem;
  border-bottom: 1px solid transparent;
  cursor: move;
}

footer {
  position: absolute;
  bottom: 10px;
  margin-top: 2rem;
  display: flex;
  justify-content: space-evenly;
  max-width: 500px;
  width: 100vw;
}

footer > button {
  height: 6rem;
  width: 6rem;
  font-size: 3rem;
  border: 1px solid grey;
  border-radius: 50%;
  line-height: 1.5;
  padding: 10px;
  cursor: pointer;
  user-select: none;
}

button.active {
  background-color: #ffeda3;
}
</style>
