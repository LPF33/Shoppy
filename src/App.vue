<template>
  <div id="view">
    <router-view />
  </div>
  <Navbar />
</template>

<script>
import Navbar from "@/components/Navbar.vue";

export default {
  components: { Navbar },
  mounted() {
    this.$store.dispatch("checkAuth");
  },
  unmounted() {
    // safe is safe ...
    this.$store.commit("unsubscribeFirebase");
  },
};
</script>

<style>
* {
  font-family: Avenir, Helvetica, Arial, sans-serif;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
}

#app {
  display: flex;
  flex-direction: column;
  width: 100vw;
  height: 100vh;
}

#view {
  flex-grow: 1;
}

.error {
  background-color: red;
  text-align: center;
  font-size: 1.5rem;
  padding: 10px;
}
</style>
