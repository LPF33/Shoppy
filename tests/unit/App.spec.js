import { mount, shallowMount } from "@vue/test-utils";
import { createRouter, createWebHistory } from "vue-router";
import { routes } from "@/router/index.js";
import App from "@/App";
import Login from "@/views/Login";
import Main from "@/views/Main";
import Shoppy from "@/views/Shoppy";
import Cooky from "@/views/Cooky";
import Jambo from "@/views/Jambo";
import Navbar from "@/components/Navbar";

function createSetup(authenticated = false) {
  const $store = {
    state: {
      isAuthenticated: authenticated,
    },
    commit: jest.fn(),
    dispatch: jest.fn(),
    watch: jest.fn(),
  };

  const router = createRouter({
    history: createWebHistory(),
    routes,
  });

  router.beforeEach((to, from, next) => {
    if (to.matched.some((record) => record.meta.requiresAuth)) {
      if ($store.state.isAuthenticated) {
        next();
        return;
      }
      next("/login");
    } else {
      next();
    }
  });

  router.beforeEach((to, from, next) => {
    if (to.matched.some((record) => record.meta.guest)) {
      if ($store.state.isAuthenticated) {
        next("/");
        return;
      }
      next();
    } else {
      next();
    }
  });

  return { router, $store };
}

describe("Testing the App component", () => {
  it("App is mounted correctly when user is not logged in", async () => {
    const { router, $store } = createSetup();
    router.push("/");
    await router.isReady();

    const wrapper = mount(App, {
      global: {
        plugins: [router],
        mocks: { $store },
      },
    });

    expect($store.dispatch).toHaveBeenCalledTimes(1);
    expect($store.dispatch.mock.calls[0][0]).toBe("checkAuth");
    expect(wrapper.findComponent(Navbar).exists()).toBeTruthy();
    expect(wrapper.vm.$route.path).toBe("/login");
    expect(wrapper.findComponent(Login).exists()).toBeTruthy();
    expect(wrapper.findComponent(Main).exists()).toBeFalsy();
  });

  it("App is mounted correctly when user is logged in", async () => {
    const { router, $store } = createSetup(true);
    router.push("/");
    await router.isReady();

    const wrapper = mount(App, {
      global: {
        plugins: [router],
        mocks: { $store },
      },
    });

    expect($store.dispatch).toHaveBeenCalledTimes(1);
    expect(wrapper.findComponent(Navbar).exists()).toBeTruthy();
    expect(wrapper.vm.$route.path).toBe("/");
    expect(wrapper.findComponent(Main).exists()).toBeTruthy();
    expect(wrapper.findComponent(Login).exists()).toBeFalsy();
  });

  it("When user not logged in, trying to go to '/' redirected to '/login'", async () => {
    const { router, $store } = createSetup(false);
    router.push("/");
    await router.isReady();

    const wrapper = shallowMount(App, {
      global: {
        plugins: [router],
        mocks: { $store },
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.vm.$route.path).toBe("/login");
  });

  it("When user logged in, trying to go to '/login' redirected to '/'", async () => {
    const { router, $store } = createSetup(true);
    router.push("/login");
    await router.isReady();

    const wrapper = shallowMount(App, {
      global: {
        plugins: [router],
        mocks: { $store },
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.vm.$route.path).toBe("/");
  });

  it("When user logged in, navigating to '/shoppy' will render Shoppy component", async () => {
    const { router, $store } = createSetup(true);
    router.push("/shoppy");
    await router.isReady();

    const wrapper = shallowMount(App, {
      global: {
        plugins: [router],
        mocks: { $store },
      },
    });

    expect(wrapper.vm.$route.path).toBe("/shoppy");
    expect(wrapper.findComponent(Shoppy).exists).toBeTruthy();
  });

  it("When user logged in, navigating to '/cooky' will render Cooky component", async () => {
    const { router, $store } = createSetup(true);
    router.push("/cooky");
    await router.isReady();

    const wrapper = shallowMount(App, {
      global: {
        plugins: [router],
        mocks: { $store },
      },
    });

    expect(wrapper.vm.$route.path).toBe("/cooky");
    expect(wrapper.findComponent(Cooky).exists).toBeTruthy();
  });

  it("When user logged in, navigating to '/jambo' will render Jambo component", async () => {
    const { router, $store } = createSetup(true);
    router.push("/jambo");
    await router.isReady();

    const wrapper = shallowMount(App, {
      global: {
        plugins: [router],
        mocks: { $store },
      },
    });

    expect(wrapper.vm.$route.path).toBe("/jambo");
    expect(wrapper.findComponent(Jambo).exists).toBeTruthy();
  });
});
