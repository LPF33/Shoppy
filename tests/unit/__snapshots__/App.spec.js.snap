// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the App component When user logged in, trying to go to '/login' redirected to '/' 1`] = `
<div id="view">
  <router-view-stub name="default"></router-view-stub>
</div>
<navbar-stub></navbar-stub>
`;

exports[`Testing the App component When user not logged in, trying to go to '/' redirected to '/login' 1`] = `
<div id="view">
  <router-view-stub name="default"></router-view-stub>
</div>
<navbar-stub></navbar-stub>
`;
